File: src/frontend/vue-notus-main/vue-notus-main/vue-notus-main/src/service/ChatLieuService.js
import api from './api';
import ChatLieu from '../models/ChatLieu';

class ChatLieuService {
  async getAll() {
    const response = await api.get('/chat-lieu');
    return response.data.map(item => new ChatLieu(item));
  }

  async getById(id) {
    const response = await api.get(`/chat-lieu/${id}`);
    return new ChatLieu(response.data);
  }

  async create(chatLieu) {
    const response = await api.post('/chat-lieu', chatLieu.toJSON());
    return new ChatLieu(response.data);
  }

  async update(id, chatLieu) {
    const response = await api.put(`/chat-lieu/${id}`, chatLieu.toJSON());
    return new ChatLieu(response.data);
  }

  async delete(id) {
    await api.delete(`/chat-lieu/${id}`);
  }

  async search(keyword) {
    const response = await api.get('/chat-lieu/search', {
      params: { keyword }
    });
    return response.data.map(item => new ChatLieu(item));
  }
}

export default new ChatLieuService();

File: src/main/java/org/example/datnbbook/service/ChatLieuService.java
package org.example.datnbbook.service;

import org.example.datnbbook.dto.ChatLieuDTO;
import org.example.datnbbook.model.ChatLieu;
import org.example.datnbbook.model.ChatLieu;
import org.example.datnbbook.repository.ChatLieuRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class ChatLieuService {
    @Autowired
    private ChatLieuRepository chatLieuRepository;


    public List<ChatLieu> getAll() {
        return chatLieuRepository.findAllByDeletedFalseOrderByIdDesc();
    }

    public ChatLieu getById(Integer id) {
        return chatLieuRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Chất liệu không tồn tại với id: " + id));
    }

    public ChatLieu create(ChatLieu sp) {
        if (sp.getMaChatLieu() == null || sp.getMaChatLieu().isEmpty()) {
            String nextMaChatLieu = chatLieuRepository.getNextSequenceValue();
            sp.setMaChatLieu(nextMaChatLieu);
        }
        sp.setDeleted(false); // Đảm bảo Chất liệu mới không bị đánh dấu xóa
        return chatLieuRepository.save(sp);
    }

    public ChatLieuDTO update(Integer id, ChatLieuDTO chatLieuDTO) {
        ChatLieu chatLieu = chatLieuRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Chất liệu không tồn tại với id: " + id));

        BeanUtils.copyProperties(chatLieuDTO, chatLieu ,"maChatLieu","id", "deleted");
        chatLieu = chatLieuRepository.save(chatLieu);
        return convertToDTO(chatLieu);
    }

    public void delete(Integer id) {
        ChatLieu chatLieu = chatLieuRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Chất liệu không tồn tại với id: " + id));
        chatLieu.setDeleted(true);
        chatLieuRepository.save(chatLieu);
    }

    public List<ChatLieu> search(String keyword) {
        return chatLieuRepository.search(keyword);
    }

    private ChatLieuDTO convertToDTO(ChatLieu chatLieu) {
        ChatLieuDTO chatLieuDTO = new ChatLieuDTO();
        BeanUtils.copyProperties(chatLieu, chatLieuDTO);
        return chatLieuDTO;
    }
}
